### PR TITLE
Expose metrics to prometheus

### PR DESCRIPTION
**Pull request by tapih**
https://github.com/cybozu-go/neco/issues/747

### CODE REVIEW

File: docs/metrics.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Metrics
+=======
+
+CKE exposes the following metrics with the Prometheus format. The listen address can be configured by the CLI flag (see [here](cke.md#Usage)). All these metrics are prefixed with `cke_`
+
+| Name              | Description                                      | Type  | Labels                             |
+| ----------------- | ------------------------------------------------ | ----- | ---------------------------------- |
+| operation_running | True(=1) if any operations are running.          | Gauge |                                    |
+| boot_leader       | True(=1) if the boot server is the leader of CKE | Gauge |                                    |
```
Comment (ymmt2005) [inline docs/metrics.md]: `boot_` looks extraneous.

File: docs/metrics.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+| node_info         | The Control Plane and Worker info                | Gauge | address, rack, role, control_plane |
```
Comment (ymmt2005) [inline docs/metrics.md]: This is completely not a metric.

File: metrics.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	storage *Storage
+}
+
+// NewCollector returns a new Collector.
+func NewCollector(client *v3.Client) *Collector {
+	return &Collector{
+		metrics: map[string]Metric{
+			"operation_phase": {
+				collectors:  []prometheus.Collector{operationPhase, operationPhaseTimestampSeconds},
+				isAvailable: alwaysAvailable,
```
Comment (ymmt2005) [inline metrics.go]: Really?  I think this should only be available when the instance is the leader.

File: metrics.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+)
+
+type logger struct{}
+
+func (l logger) Println(v ...interface{}) {
+	log.Error(fmt.Sprint(v...), nil)
+}
+
+const (
+	namespace     = "cke"
```
Comment (ymmt2005) [inline metrics.go]: The name `namespace` is too generic for the package `cke`.
Consider creating `metrics` package and move this file under it.
Splitting this file should also be considered as this file is a bit large.

File: metrics.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+	scrapeTimeout = time.Second * 10
```
Comment (ymmt2005) [inline metrics.go]: The default scrape timeout value of Prometheus is also 10 seconds.
Considering that, this value should be shortened, shouldn't this?

File: metrics.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+)
+
+// Metric represents collector and availability of metric.
+type Metric struct {
```
Comment (ymmt2005) [inline metrics.go]: I think `MetricGroup` is more appropriate.

File: metrics.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+	collectors  []prometheus.Collector
+	isAvailable func(context.Context, *Storage) (bool, error)
+}
+
+// Collector is a metrics collector for CKE.
+type Collector struct {
```
Comment (ymmt2005) [inline metrics.go]: Why should this be exported?

File: metrics.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+	metrics map[string]Metric
+	storage *Storage
+}
+
+// NewCollector returns a new Collector.
+func NewCollector(client *v3.Client) *Collector {
```
Comment (ymmt2005) [inline metrics.go]: Can `*Collector` be replaced with `prometheus.Collector`?

File: pkg/cke/main.go [hunk 9]
```diff
@@ -98,9 +99,25 @@ func main() {
...
+	mux.Handle("/metrics", metricsHandler)
+	ms := &well.HTTPServer{
+		Server: &http.Server{
+			Addr:    *flgMetrics,
+			Handler: mux,
+		},
+	}
+	ms.ListenAndServe()
+
+	// API server
```
Comment (ymmt2005) [inline pkg/cke/main.go]: Why don't you merge the two?

File: storage.go [hunk 10]
```diff
@@ -490,7 +501,7 @@ func (s Storage) GetLeaderHostname(ctx context.Context) (string, error) {
 	}
 
 	if len(resp.Kvs) == 0 {
-		return "", errors.New("no leader")
```
Comment (ymmt2005) [inline storage.go]: Why do you change this line?
`ErrLeaderNotExist` is used only here.  It seems not worthy of a constant.

File: metrics.go (hunk 3 above)
Comment (morimoto-cybozu) [reply metrics.go]: That's right.  I'm updating the code and adding the following note to the doc.

> All metrics but `leader` are available only when the server is the leader of CKE.

File: metrics.go (hunk 5 above)
Comment (morimoto-cybozu) [reply metrics.go]: Absolutely, it should be shortened.
The purpose I applied this timeout was to reclaim resources as soon as possible *after* the scraping timeout.  Now I'm reminded that the collector should return as many metrics as possible even when some collection goroutine is delayed.  The shorter timeout will enable partial responses.

File: metrics.go (hunk 7 above)
Comment (morimoto-cybozu) [reply metrics.go]: There is no need to export.  Fixed.

File: storage.go (hunk 10 above)
Comment (morimoto-cybozu) [reply storage.go]: `ErrLeaderNotExist` was once used in multiple code portions, which have been removed now.  Reverted.

File: pkg/cke/main.go (hunk 9 above)
Comment (morimoto-cybozu) [reply pkg/cke/main.go]: I supposed that the REST API of CKE was too complicated to be multiplexed with `/metrics`.  Actually CKE has just two endpoints.  Merged.

File: docs/metrics.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+Metrics
+=======
+
+CKE exposes the following metrics with the Prometheus format. The listen address can be configured by the CLI flag (see [here](cke.md#Usage)). The path is fixed to `/metrics`. All these metrics are prefixed with `cke_`
```
Comment (ymmt2005) [inline docs/metrics.md]:
```suggestion
CKE exposes the following metrics with the Prometheus format at `/metrics` REST API endpoint.  All these metrics are prefixed with `cke_`
```

**Comment by ymmt2005** [comment]
@morimoto-cybozu Could you please resolve the conflict?

**Comment by morimoto-cybozu** [comment]
@ymmt2005 Yeah, sure!

**Pull request merged**